packages/itwinui-react/src/core/Dialog/Dialog.tsx

```diff
@@ -37,10 +46,34 @@ const DialogComponent = React.forwardRef((props, ref) => {
     placement,
     className,
     portal = false,
+    renderWrapperWhenClosed = true,
     ...rest
   } = props;
 
+  // When user can/should see the dialog wrapper's children
+  // (e.g. when opacity!=0 and not necessarily when isOpen=false (since dialog is visible when animating out))
+  const [isDialogWrapperChildrenVisible, setIsDialogWrapperChildrenVisible] =
```

Comment:
Open to suggestions for variable names or even different approaches to achieving the same functionality.

Reply:
I had played around with it and `mountOnEnter` before but couldn't get it working. Upon retrying it now, I was able to get it working after a few more modifications.

So now the logic is slightly simplified (https://github.com/iTwin/iTwinUI/pull/1865/commits/3ccbc9bb341ce4957eb339b307ab7e255519c4c8).

Reply:
So you weren't able to remove `isDialogWrapperChildrenVisible` and the internal `isOpen` state?

Triggering `useEffect` based on props is generally unnecessary and [explicitly discouraged by React](https://react.dev/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state).

Reply:
> Triggering `useEffect` based on props is generally unnecessary and [explicitly discouraged by React](https://react.dev/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state).

I believe their recommendation is to avoid a stale pass with an incorrect value. But in our case, we intentionally want the first pass to happen with `isOpen=false` and the next passes with `isOpen=true`. This is for proper CSS transitions. So, I don't think this React recommendation applies in our case.

I still tried to remove it (https://github.com/iTwin/iTwinUI/pull/1865/commits/00d7a38ccc8105f6c05bdce666133a53329886b9). But I needed `useIsFirstRender()` from `ShadowRoot.tsx`. I also needed to add `queueMicrotask` to make it work in Dialog.

I'm not sure if `queueMicrotask` would regress something in ShadowRoot. Also, not sure if `useIsFirstRender` is an apt name when using `queueMicrotask` that may not run immediately after the first render and before the second.

> So you weren't able to remove `isDialogWrapperChildrenVisible` and the internal `isOpen` state?

I wasn't able to remove `isDialogWrapperChildrenVisible` since I need a way to set `isOpen` in the context to `true` only after the wrapper has been rendered/mounted.

If we really want to avoid a new state variable, I can look into creating a new component for the wrapper. I can then use `isFirstRender` in that new component to set `isOpen=false` when `isFirstRender` and `isOpen=true` when `!isFirstRender`. This is just an idea, not sure if it's feasible or even better than the current approach.

Reply:
If context is the only problem, then it could be moved down underneath `Transition`.

```jsx
<Portal …>
  <Transition …>
    <Box …>
      <DialogContext.Provider …>
        {children}
      </DialogContext.Provider>
    </Box>
  </Transition>
</Portal>
```

A new component (within the same file) also makes sense if this doesn't work.

Basically, updating state inside `useEffect` should be the last resort and should have a very clear and strong reason behind it (which I'm not seeing in this case). The whole `isFirstRender` thing is a hack for server-side rendering declarative shadow dom, so it shouldn't be used here.

Reply:
I am not sure of how to achieve the same logic without a useEffect + state variable and also without `useIsFirstRender`.

I created an inner helper function that has a `useEffect` + stateVariable (https://github.com/iTwin/iTwinUI/pull/1865/commits/d80f2f13e0e3308426d5a8ca77e8a6de3528b83b). Although it has a useEffect that updates a state variable, at least this time the dependency array is empty. So, it runs only once. I think updating states in such `useEffect`s that only run once are alright.

Reply:
During internal discussion/pairing session, we were able to get rid of the `useEffect` by using a CSS animation. 👍